[PM-4154] Introduce Bulk Encrypt Service for Faster Unlock Times

quexten opened the pull request:
## Type of change

<!-- (mark with an `X`) -->

```
- [ ] Bug fix
- [x] New feature development
- [ ] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective (Updated)

Introduce a `BulkEncryptionService` that uses multiple workers in order to decrypt items during unlock faster. This only applies to bulk decryption, and single decryption will stay in the EncryptService in the foreground thread. This feature is featureflagged behind PM-4154-bulk-encryption-service (https://github.com/bitwarden/server/pull/4091).
Much of the code is duplicated, so that toggling off will not change behavior compared to before, since bulk decryption is an operation that we really do not want failing. The duplicate/redundant classes and functions have been marked deprecated and shall be removed once the feature is rolled out and tested properly.

Tested on web/browser(mv3,mv2)/desktop, but desktop will not use multi-worker decryption for now (PM-6797).

## Code changes


## Screenshots

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team
- Ensure that all UI additions follow [WCAG AA requirements](https://contributing.bitwarden.com/contributing/accessibility/)

quexten commented:
Partially fixes: https://github.com/bitwarden/clients/issues/1597

eliykat reviewed (commented):
I like it!

A few questions/comments:

1. at the moment, we take 1 thread with a TTL of 3 minutes. This was not a particularly disciplined approach (by me) because 1 inactive thread doesn't tie up many system resources. But if we're taking the max threads available to us, I assume we need to be more aggressive about freeing up system resources once we're done. I don't have any firm ideas here, can we just spin them up & terminate them every time the method is called? Or shoudl we have a shorter TTL which more accurately reflects inactivity? (at the moment the TTL starts counting when work is sent to the worker, not when the worker becomes inactive, it's not particularly robust)

2. I suggest we define a minimum number of items per worker, otherwise (for example) decrypting a small vault of 10 items could take 10 separate threads. This is additional overhead for no user benefit. I don't know what this number would be.

3. Can you please wrap your changes behind a [feature flag](https://contributing.bitwarden.com/contributing/feature-flags). For example:

```ts
// feature-flag.enum.ts
MultiWorkerDecryption = "multi-worker-decryption",

// decryptItems
const multiWorkerDecryptionEnabled = await firstValueFrom(this.configService.getFeatureFlag$(FeatureFlag.MultiWorkerDecryption));
if (multiWorkerDecryptionEnabled) {
  return decryptItems_vNext(...)
} else {
  return decryptItems_Legacy(...)
}  
```

Refer to that documentation for more info about how to use these. This will let us put it into production more safely.

quexten commented:
Hmm, so I can't inject the configservice to the encrypt service, because it depends on the authservice, which depends on the cryptoservice, which depends on the encrypt service, causing a dependency cycle. I don't see a good way to break this dependency cycle.

For the previous background thread "multithread-encrypt" service feature and for cipherkey encryption, it was done via this facility instead: https://github.com/bitwarden/clients/blob/6b7edced8e66b23855bea2d478536f83931a54cf/libs/common/src/platform/misc/flags.ts#L3-L7

Though this seems to be more for build-time configuration, not run-time server toggleable features. So this might not be optimal for rolling back in production?

Any advice on which path to take forward here?

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/bitwarden/clients?pullRequest=6465) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/bitwarden/clients?pullRequest=6465) before we can accept your contribution.<br/>**18** out of **25** committers have signed the CLA.<br/><br/>:white_check_mark: justindbaur<br/>:white_check_mark: aj-rosado<br/>:white_check_mark: kspearrin<br/>:white_check_mark: willmartian<br/>:white_check_mark: Jingo88<br/>:white_check_mark: cagonzalezcs<br/>:white_check_mark: rr-bw<br/>:white_check_mark: dani-garcia<br/>:white_check_mark: MGibson1<br/>:white_check_mark: vgrassia<br/>:white_check_mark: Hinton<br/>:white_check_mark: eliykat<br/>:white_check_mark: vincentsalucci<br/>:white_check_mark: urbinaalex17<br/>:white_check_mark: abergs<br/>:white_check_mark: ike-kottlowski<br/>:white_check_mark: quexten<br/>:white_check_mark: porzana<br/>:x: JaredSnider-Bitwarden<br/>:x: amorask-bitwarden<br/>:x: trmartin4<br/>:x: jlf0dev<br/>:x: coroiu<br/>:x: shane-melton<br/>:x: gbubemismith<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/bitwarden/clients?pullRequest=6465) it.</sub>

quexten commented:
Uhh, sorry about the pings, messed up a rebase there and the bots picked it up and marked it for review.

quexten commented:
>  at the moment, we take 1 thread with a TTL of 3 minutes. This was not a particularly disciplined approach (by me) because 1 inactive thread doesn't tie up many system resources. But if we're taking the max threads available to us, I assume we need to be more aggressive about freeing up system resources once we're done. I don't have any firm ideas here, can we just spin them up & terminate them every time the method is called? Or shoudl we have a shorter TTL which more accurately reflects inactivity? (at the moment the TTL starts counting when work is sent to the worker, not when the worker becomes inactive, it's not particularly robust)

Changed ttl to 1 minute, but changed the behaviour so that the ttl is disabled during decryption. Also added a limit of min(system threads, 16) to the number of workers, and a minimum of 500 items.

eliykat commented:
> Hmm, so I can't inject the configservice to the encrypt service, because it depends on the authservice, which depends on the cryptoservice, which depends on the encrypt service, causing a dependency cycle. I don't see a good way to break this dependency cycle.

There is an upcoming refactor that might help here, I'll check in with the team responsible. Otherwise we could evaluate the flag in `CipherService` and pass the boolean as an argument when calling `decryptItems`, but I'd prefer not to if we can avoid it.

> For the previous background thread "multithread-encrypt" service feature and for cipherkey encryption, it was done via this facility instead:
> 
> https://github.com/bitwarden/clients/blob/6b7edced8e66b23855bea2d478536f83931a54cf/libs/common/src/platform/misc/flags.ts#L3-L7
> 
> Though this seems to be more for build-time configuration, not run-time server toggleable features. So this might not be optimal for rolling back in production?

You've understood correctly, that is no longer used and should be cleaned up.

quexten commented:
> There is an upcoming refactor that might help here, I'll check in with the team responsible. Otherwise we could evaluate the flag in CipherService and pass the boolean as an argument when calling decryptItems, but I'd prefer not to if we can avoid it.

Is there a public PR for this change that I can keep an eye on at the moment or this internally tracked work for now?

If the refactor is a more long-term initiative, I'm happy to change it to pass this flag as an argument for now with a note to remove it as soon as is viable.

quexten commented:
To be honest, I'd like to avoid adding user configurable options unless absolutely needed.

I'd rather in the next step make a decision on good values based on empirical data.
Next, there are other, further gains that can be had in decryption, that, combined with this change do not add user configuration complexity, but make the decryption time so fast that it's not noticeable.

ccben87 commented:
@eliykat Is there somewhere we can track the previously mentioned upcoming refactor? I'm really keen to see this get implemented as the performance impact we are seeing with over 11k ciphers is significantly impacting us.

eliykat commented:
Thanks for your patience, @quexten our circular dependency issues should be resolved by #8279.

@ccben87 please note that this code is only used by the web vault and browser extension clients, it's not (yet?) used by the desktop client and is totally separate from mobile. Just FYI in case you're using one of those other clients.

ccben87 commented:
@eliykat Thanks for replying! We mostly use the web vault and browser extension. However, I think some may use desktop client. Where is the code for the desktop client or what do you mean by not yet used by desktop client? Will it be used by desktop client at some point?

quexten commented:
> @eliykat Thanks for replying! We mostly use the web vault and browser extension. However, I think some may use desktop client. Where is the code for the desktop client or what do you mean by not yet used by desktop client? Will it be used by desktop client at some point?

While I have not investigated the relevant code yet, the multi-thread (background thread) decryption service, which this PR extends, is disabled at build time in the desktop client:
https://github.com/bitwarden/clients/blob/bcb2a976b094f57f1f7e1261e2692f12103d7b16/apps/desktop/config/base.json#L4

eliykat commented:
@ccben87 The desktop client has a race condition when quickly switching between large accounts - its state can get out of sync between accounts because it doesn't force you to wait for the account switch to fully settle. e.g. if you change accounts rapidly, it might try to decrypt the new vault items with the old encryption key. It just so happens that this isn't a problem if you're not using multithreading, because decryption will lock up the UI 😅

Browser avoids this because its implementation is more robust, it forces you to wait until the selected account is "ready"/settled before it lets you interact with the UI again. So we need to implement a similar thing for desktop, but I don't believe that's currently scheduled for work.

eliykat commented:
Update to my previous comment - I've followed up internally and the team in charge of account switching is going to schedule some work for the Desktop client to address this, which should then allow us to enable multithread decryption there as well. (No ETA yet but hopefully soon.)

quexten commented:
> Thanks for your patience, @quexten our circular dependency issues should be resolved by #8279.
> 
> @ccben87 please note that this code is only used by the web vault and browser extension clients, it's not (yet?) used by the desktop client and is totally separate from mobile. Just FYI in case you're using one of those other clients.

For any other community readers not keeping close eyes on related PRs, instead of #8279 , this is now waiting on #8325 

For the desktop multi-threading, I don't think there is anything that this PR needs to do, so I don't think that's a blocker?

Regardless, will re-base and add the feature flag once #8325 is merged.

ccben87 commented:
@quexten Is this still waiting on you or is it ready to be merged? If there's still work to be done, I could take a look at it but I'm new to Typescript.

quexten commented:
@ccben87 I plan to get back to this PR as soon as I find time, though that is probably in middle of May.

quexten commented:
On initial check, there is a new/different dependency cycle now. ConfigApiService needs the tokenservice, which needs the EncryptService which needs the ConfigApiService.

quexten commented:
Refactoring to move `decryptItems` to a new `BulkEncryptService`

quexten commented:
Changed to "Bulk Encrypt Service" now and gated the feature behind a feature flag. The plan is to remove `decryptItems` from EncryptService,` multithread-encrypt.service.implementation.ts`, `BrowserFallbackEncryptServiceImplementation`, `BrowserMultithreadEncryptServiceImplementation` entirely, once the feature has been rolled out and tested, leading to an overall less complex implementation. Right now there is a lot of duplication between the old code and new code, since I'd rather be *very* defensive when rolling out a feature on vault decryption.

quexten commented:
Once #9077 is merged, this PR can be further shrunk.

willmartian commented:
@quexten have you done any perf analysis on this? I am curious what gains we might see at various vault sizes.

eliykat commented:
@MGibson1 originally asked me to review this as a community PR because I was familiar with the multithread stuff. However, the recurring blocker here has been circular dependencies between our services, which is impacted by Platform's work and vision for our services structure. Now that this PR is making further changes to encryption-related services, I'd like to kick it back to Platform (@justindbaur) so that they can review it directly and make sure they're happy with it.

quexten commented:
@willmartian While not extensive (just one run using performance.now, for a 20k cipher/80k cipherstring vault, in Firefox) the results are fairly clear. On an Apple M3 Pro: Single worker: 4.5s, bulk service: 1.3s. On an Intel i7-1185G7, on power-saver: ~21s to 8.5s, without powersaver 9s to 3.3s. 

I can get more accurate measurements, but I think the benefit is pretty clear from these alone.
This is just for the call to decryptItems. Whole unlock time ofcourse also includes KDF time, state parsing, and search indexing.

quexten commented:
Thanks for the update @eliykat, since #9077 was merged, I have now removed the factory methods, bringing this PR to a more reasonable size.

willmartian commented:
> I can get more accurate measurements

@quexten No need, just wanted a rough picture. This is great, thank you very much!

justindbaur commented on libs/common/src/platform/services/cryptography/fallback-bulk-encrypt.service.implementation.ts:
```diff
@@ -0,0 +1,23 @@
+/**
+ * @deprecated For the feature flag from PM-4154, remove once feature is rolled out
+ */
+export class FallbackBulkEncryptServiceImplementation implements BulkEncryptService {
```
nit: You can remove the `Implementation` suffix since it already has a unique prefix.

quexten replied:
Fixed in 9793d5768d3ee99402f79589fcf0b253df7a4335, thanks!

justindbaur commented on apps/browser/src/background/main.background.ts:
```diff
@@ -1167,6 +1173,13 @@ export default class MainBackground {
     this.webRequestBackground?.startListening();
     this.syncServiceListener?.startListening();
 
+    if (await this.configService.getFeatureFlag(FeatureFlag.PM4154_BulkEncryptionService)) {
+      this.bulkEncryptService = new BulkEncryptServiceImplementation(
+        this.cryptoFunctionService,
+        this.logService,
+      );
+    }
```
Since all the services have been constructed will the `CipherService` actually get this? I know if you don't assign it to anything and some service injects it then it will just get `undefined`. Since you do set it to a reference and then update that references I'm inclined to think it will work, just making sure.

I don't think we will be able to use this service in the MV3 service worker though. So I think you'll need to add this check: `BrowserApi.isManifestVersion(2)`.

quexten replied:
Not sure how this made it past my local testing. Fixed this in 9793d5768d3ee99402f79589fcf0b253df7a4335 by injecting the BulkEncryptService into the FallbackEncryptService reference at runtime, this way the reference we provided to the cipher service (and transitively to many other services) does not need to be replaced.
After the feature flag is done we can rip out this construct.

For mv3, I also added a check, I guess we can re-implement this once the offscreen api, reverted in https://github.com/bitwarden/clients/pull/9221 is re-introduced.

justindbaur replied:
Thank you, can we just do those checks in the opposite order though so we do the simple sync check first and can avoid a possible api call if we are in MV3.

quexten replied:
That makes sense. Changed the order in [59e2ff3](https://github.com/bitwarden/clients/pull/6465/commits/59e2ff3e92ab9b2a8a2f8f715e9d2efbc3b7651d).

ccben87 commented:
@quexten is this ready for review or are you still working on it?

quexten commented:
@ccben87 From my side this is ready for review.

jlf0dev commented on libs/angular/src/services/jslib-services.module.ts:
```diff
@@ -1178,6 +1189,9 @@ const safeProviders: SafeProvider[] = [
   }),
 ];
 
+/**
+ * @deprecated
+ */
```
non-blocking: Can you provide some context here why these are deprecated?

jlf0dev commented on libs/common/src/platform/services/cryptography/bulk-encrypt.service.implementation.ts:
```diff
@@ -0,0 +1,159 @@
+   */
+  async decryptItems<T extends InitializerMetadata>(
+    items: Decryptable<T>[],
+    key: SymmetricCryptoKey,
```
non-blocking: This is technically Platforms code, so really up to @justindbaur, but I'd love to see us check the `key` here as well. To be fair this may not be realistic considering where this is used, so feel free to ignore if so.

justindbaur replied:
I would like that if we can make it throw when given `null`. I think I also want it to throw if `items` was null instead of returning `[]` but I see that is how it already works so it's fine to leave for now.

quexten replied:
I had a look at the corresponding code for the single threaded encrypt service implementation and it only checks for whether the key is null, so I added the same check to the bulk encrypt service. Anything more than that (comparing to private key) should be done in higher layers (when setting the key in state f.e) in my opinion.

jlf0dev commented on libs/common/src/platform/services/cryptography/bulk-encrypt.service.implementation.ts:
```diff
@@ -0,0 +1,159 @@
+   * Sends items to a set of web workers to decrypt them. This utilizes multiple workers to decrypt items
+   * faster without interrupting other operations (e.g. updating UI).
+   */
+  async getDecryptedItemsFromWorkers<T extends InitializerMetadata>(
```
non-blocking: does this need to be public?

quexten replied:
Changed this to be private

jlf0dev commented on libs/common/src/platform/services/cryptography/bulk-encrypt.service.implementation.ts:
```diff
@@ -0,0 +1,159 @@
+
+    this.clearTimeout();
+
+    let numberOfWorkers = Math.min(navigator.hardwareConcurrency, maxWorkers);
```
question: `navigator.hardwareConcurrency` doesn't seem to have safari support yet. Is that going to be an issue? Should we just revert to encrypt service on that platform?

https://developer.mozilla.org/en-US/docs/Web/API/Navigator/hardwareConcurrency

quexten replied:
I think this document might be out of date. In my safari installation, navigator.hardwareConcurrency works fine (and returns a fixed 8, which is less than the cores of my CPU). But we can add a check that defaults to 1 just in case.

quexten replied:
Defaults to 1 now

jlf0dev commented on libs/common/src/platform/services/cryptography/bulk-encrypt.service.implementation.ts:
```diff
@@ -0,0 +1,159 @@
+      `Finished decrypting ${decryptedItems.length} items using ${numberOfWorkers} workers`,
+    );
+
+    this.restartTimeout();
```
question: i'm probably just missing this, but how does the timeout here work if it comes after we await the results? Don't we need to start a timer before awaiting the results?

quexten replied:
There is an explicit change in behaviour in this implementation compared to the prior single-threaded background thread implementation regarding timeout. The idea is to never timeout if bulk decryption takes too long (as that might just be a slow cpu). The reason for the timeout, is to "spin down" _unused_ worker threads, so they do not use up resources in the background.

quexten replied:
This was @Thomas-Avery's initial comment on the TTL:
> at the moment, we take 1 thread with a TTL of 3 minutes. This was not a particularly disciplined approach (by me) because 1 inactive thread doesn't tie up many system resources. But if we're taking the max threads available to us, I assume we need to be more aggressive about freeing up system resources once we're done. I don't have any firm ideas here, can we just spin them up & terminate them every time the method is called? Or shoudl we have a shorter TTL which more accurately reflects inactivity? (at the moment the TTL starts counting when work is sent to the worker, not when the worker becomes inactive, it's not particularly robust)

jlf0dev reviewed (commented):
Super cool stuff, I had a couple questions but everything else was non-blocking.

shane-melton reviewed (approved):
I know there's still ongoing discussion and will possibly need re-approval, but the changes to Vault owned files look good to go!

justindbaur reviewed (commented):
Looks pretty good from my end, just the one suggestion from @jlf0dev, which I also consider non-blocking. 

But it also looks like there are some test and CLI build problems.

quexten commented:
Added a check for the key being null (that's what the single threaded encrypt service does) and fixed the cli build.

jlf0dev reviewed (approved):
Glad to see this come together 👏  looks good from my end!

quexten commented:
Had to resolve merge conflicts, so this needs another approval. Enabling auto-merge, since this is feature flagged, so QA can test on main.
